pkg/plugins/nop/sender.go

```diff
@@ -0,0 +1,146 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nop
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/xmidt-org/ears/internal/pkg/rtsemconv"
+	"github.com/xmidt-org/ears/pkg/secret"
+	"github.com/xmidt-org/ears/pkg/tenant"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/unit"
+	"time"
+
+	"github.com/goccy/go-yaml"
+	"github.com/xmidt-org/ears/pkg/event"
+	pkgplugin "github.com/xmidt-org/ears/pkg/plugin"
+	"github.com/xmidt-org/ears/pkg/sender"
+)
+
+func NewSender(tid tenant.Id, plugin string, name string, config interface{}, secrets secret.Vault) (sender.Sender, error) {
+	var cfg SenderConfig
+	var err error
+	switch c := config.(type) {
+	case string:
+		err = yaml.Unmarshal([]byte(c), &cfg)
+	case []byte:
+		err = yaml.Unmarshal(c, &cfg)
+	case SenderConfig:
+		cfg = c
+	case *SenderConfig:
+		cfg = *c
+	}
+	if err != nil {
+		return nil, &pkgplugin.InvalidConfigError{
+			Err: err,
+		}
+	}
+	cfg = cfg.WithDefaults()
+	err = cfg.Validate()
+	if err != nil {
+		return nil, err
+	}
+	s := &Sender{
+		config: cfg,
+		name:   name,
+		plugin: plugin,
+		tid:    tid,
+	}
+	// metric recorders
+	meter := global.Meter(rtsemconv.EARSMeterName)
+	commonLabels := []attribute.KeyValue{
+		attribute.String(rtsemconv.EARSPluginTypeLabel, rtsemconv.EARSPluginTypeNopSender),
+		attribute.String(rtsemconv.EARSPluginNameLabel, s.Name()),
+		attribute.String(rtsemconv.EARSAppIdLabel, s.tid.AppId),
+		attribute.String(rtsemconv.EARSOrgIdLabel, s.tid.OrgId),
+	}
+	s.eventSuccessCounter = metric.Must(meter).
+		NewInt64Counter(
+			rtsemconv.EARSMetricEventSuccess,
+			metric.WithDescription("measures the number of successful events"),
+		).Bind(commonLabels...)
+	s.eventFailureCounter = metric.Must(meter).
+		NewInt64Counter(
+			rtsemconv.EARSMetricEventFailure,
+			metric.WithDescription("measures the number of unsuccessful events"),
+		).Bind(commonLabels...)
+	s.eventBytesCounter = metric.Must(meter).
+		NewInt64Counter(
+			rtsemconv.EARSMetricEventBytes,
+			metric.WithDescription("measures the number of event bytes processed"),
+			metric.WithUnit(unit.Bytes),
+		).Bind(commonLabels...)
+	s.eventProcessingTime = metric.Must(meter).
+		NewInt64Histogram(
+			rtsemconv.EARSMetricEventProcessingTime,
+			metric.WithDescription("measures the time an event spends in ears"),
+			metric.WithUnit(unit.Milliseconds),
+		).Bind(commonLabels...)
+	s.eventSendOutTime = metric.Must(meter).
+		NewInt64Histogram(
+			rtsemconv.EARSMetricEventSendOutTime,
+			metric.WithDescription("measures the time ears spends to send an event to a downstream data sink"),
+			metric.WithUnit(unit.Milliseconds),
+		).Bind(commonLabels...)
+	return s, nil
+}
+
+func (s *Sender) Send(e event.Event) {
+	buf, err := json.Marshal(e.Payload())
```

Comment:
is payload in json format a requirement?

Reply:
no, it doesn't even get sent, i'm just using this to measure the bytes if it were to be json serialized for metrics purposes